### PR TITLE
Add support for --optfile to gdal_calc.py

### PR DESCRIPTION
**Pull request by fitnr**
## What does this PR do?
This adds the `--optfile` option to `gdal_calc.py`.

Other Python scripts use `gdal.GeneralCmdLineProcessor`, which parses the `optfile` flag. Gdal_calc, on the other hand, uses the Python optparse module. As in the C code, lines beginning with `#` are ignored.

A new test is included.

## Tasklist

 - [X] Add test case(s)
 - [ ] Review
 - [ ] Adjust for comments
 - [ ] All CI builds and checks have passed

### CODE REVIEW

File: autotest/pyscripts/test_gdal_calc.py [hunk 1]
```diff
@@ -332,6 +332,67 @@ def test_gdal_calc_py_6():
 
     return 'success'
 
+###############################################################################
+# test --optfile
+
+def test_gdal_calc_py_7():
```
Comment (rouault) [inline autotest/pyscripts/test_gdal_calc.py]: sorry, this is not the unittest module . You have to manually add test_gal_calc_py_7 to the gdaltest_list array at the bottom of the file

**Comment by rouault** [comment]
Tests failure on multiple Travis-CI targets:
```
  TEST: test_gdal_calc_py_2 ... fail (blowup)
Traceback (most recent call last):
  File "pymod/gdaltest_python2.py", line 43, in run_func
    result = func()
  File "./test_gdal_calc.py", line 116, in test_gdal_calc_py_2
    test_py_scripts.run_py_script(script_path, 'gdal_calc', '-A tmp/test_gdal_calc_py.tif --A_band 1 -B tmp/test_gdal_calc_py.tif --B_band 2 --calc=A+B --overwrite --outfile tmp/test_gdal_calc_py_2_1.tif')
  File "../pymod/test_py_scripts.py", line 84, in run_py_script
    return run_py_script_as_py_module(script_path, script_name, concatenated_argv)
  File "../pymod/test_py_scripts.py", line 146, in run_py_script_as_py_module
    exec(script_name + '.main()')
  File "<string>", line 1, in <module>
  File "/home/travis/build/OSGeo/gdal/autotest/pyscripts/../../gdal/swig/python/scripts/gdal_calc.py", line 424, in main
    add_alpha_args(parser, sys.argv)
  File "/home/travis/build/OSGeo/gdal/autotest/pyscripts/../../gdal/swig/python/scripts/gdal_calc.py", line 408, in add_alpha_args
    given_args = set([a[1] for a in argv if a[1:2] in AlphaList] + ['A'])
IndexError: string index out of range
  TEST: test_gdal_calc_py_3 ... fail (blowup)
Traceback (most recent call last):
  File "pymod/gdaltest_python2.py", line 43, in run_func
    result = func()
  File "./test_gdal_calc.py", line 161, in test_gdal_calc_py_3
    test_py_scripts.run_py_script(script_path, 'gdal_calc', '-A tmp/test_gdal_calc_py.tif --allBands A --calc=A --overwrite --outfile tmp/test_gdal_calc_py_3.tif')
  File "../pymod/test_py_scripts.py", line 84, in run_py_script
    return run_py_script_as_py_module(script_path, script_name, concatenated_argv)
  File "../pymod/test_py_scripts.py", line 146, in run_py_script_as_py_module
    exec(script_name + '.main()')
  File "<string>", line 1, in <module>
  File "/home/travis/build/OSGeo/gdal/autotest/pyscripts/../../gdal/swig/python/scripts/gdal_calc.py", line 424, in main
    add_alpha_args(parser, sys.argv)
  File "/home/travis/build/OSGeo/gdal/autotest/pyscripts/../../gdal/swig/python/scripts/gdal_calc.py", line 408, in add_alpha_args
    given_args = set([a[1] for a in argv if a[1:2] in AlphaList] + ['A'])
IndexError: string index out of range
  TEST: test_gdal_calc_py_4 ... fail (blowup)
Traceback (most recent call last):
  File "pymod/gdaltest_python2.py", line 43, in run_func
    result = func()
  File "./test_gdal_calc.py", line 200, in test_gdal_calc_py_4
    test_py_scripts.run_py_script(script_path, 'gdal_calc', '-A tmp/test_gdal_calc_py.tif -B tmp/test_gdal_calc_py_4_1.tif --B_band 1 --allBands A --calc=A+B --NoDataValue=999 --overwrite --outfile tmp/test_gdal_calc_py_4_2.tif')
  File "../pymod/test_py_scripts.py", line 84, in run_py_script
    return run_py_script_as_py_module(script_path, script_name, concatenated_argv)
  File "../pymod/test_py_scripts.py", line 146, in run_py_script_as_py_module
    exec(script_name + '.main()')
  File "<string>", line 1, in <module>
  File "/home/travis/build/OSGeo/gdal/autotest/pyscripts/../../gdal/swig/python/scripts/gdal_calc.py", line 424, in main
    add_alpha_args(parser, sys.argv)
  File "/home/travis/build/OSGeo/gdal/autotest/pyscripts/../../gdal/swig/python/scripts/gdal_calc.py", line 408, in add_alpha_args
    given_args = set([a[1] for a in argv if a[1:2] in AlphaList] + ['A'])
IndexError: string index out of range
```

**Comment by fitnr** [comment]
This failed because I changed AlphaList to a string rather than a list, thinking I was cleaning things up. But in Python, `'' in 'ABC' == True`, while `'' in ['A', 'B', 'C'] == False`. Lesson learned: don't touch anything that isn't necessary.

**Comment by fitnr** [comment]
This rather simple change now has a lot of commits. I'd be happy to compress them into a single commit and re-submit this pull request, if that's preferred.

**Comment by rouault** [comment]
I've used the "squash" button to compress into a single commit. Thanks

**Pull request merged**